ports/esp32/esp32_rmt.c

```diff
@@ -64,51 +65,79 @@ STATIC mp_obj_t esp32_rmt_make_new(const mp_obj_type_t *type, size_t n_args, siz
         { MP_QSTR_id,        MP_ARG_REQUIRED | MP_ARG_INT, {.u_int = -1} },
         { MP_QSTR_pin,       MP_ARG_REQUIRED | MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
         { MP_QSTR_clock_div,                   MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 8} }, // 100ns resolution
-        { MP_QSTR_carrier_duty_percent,        MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 50} },
-        { MP_QSTR_carrier_freq,                MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_idle_level,                  MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} }, // 0 as default
+        { MP_QSTR_tx_carrier,                  MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
     };
     mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
     mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
     mp_uint_t channel_id = args[0].u_int;
     gpio_num_t pin_id = machine_pin_get_id(args[1].u_obj);
     mp_uint_t clock_div = args[2].u_int;
+    mp_uint_t idle_level = args[3].u_int;
+    mp_obj_t tx_carrier_obj = args[4].u_obj;
 
-    bool carrier_en = false;
-    mp_uint_t carrier_duty_percent = 0;
-    mp_uint_t carrier_freq = 0;
-
-    if (args[4].u_int > 0) {
-        carrier_en = true;
-        carrier_duty_percent = args[3].u_int;
-        carrier_freq = args[4].u_int;
-    }
+    rmt_config_t config;
 
     if (clock_div < 1 || clock_div > 255) {
         mp_raise_ValueError(MP_ERROR_TEXT("clock_div must be between 1 and 255"));
     }
 
+    if (tx_carrier_obj != mp_const_none) {
+        if (!mp_obj_is_type(tx_carrier_obj, &mp_type_tuple)) {
+            mp_raise_TypeError(MP_ERROR_TEXT("tx_carrier must be a tuple"));
+        }
+
+        mp_uint_t tx_carrier_details_length = 0;
+        mp_obj_t *tx_carrier_details = NULL;
+        mp_obj_tuple_get(tx_carrier_obj, &tx_carrier_details_length, &tx_carrier_details);
+
+        if (tx_carrier_details_length != 3 ||
+            !mp_obj_is_int(tx_carrier_details[0]) ||
+            !mp_obj_is_int(tx_carrier_details[1]) ||
+            !mp_obj_is_int(tx_carrier_details[2])) {
+            mp_raise_ValueError(MP_ERROR_TEXT("tx_carrier must be a tuple of three integers: frequency, duty and level"));
+        }
```

Comment:
If you use `mp_obj_get_array_fixed_n(tx_carrien_obj, 3, &items)` then it'll automatically give a nice error message, you won't need to raise any yourself.  This whole first section of the if-statement can be replaced by that call.